com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboardCategory.cs

```diff
@@ -0,0 +1,623 @@
+        {
+            m_DragIndicator.visible = visible;
+        }
+
+        public bool CategoryContains(List<ISelectable> selection)
+        {
+            if (selection == null)
+                return false;
+
+            // Look for at least one selected element in this category to accept drop
+            foreach (ISelectable selected in selection)
+            {
+                VisualElement selectedElement = selected as VisualElement;
+
+                if (selected != null && Contains(selectedElement))
+                {
+                    if (canAcceptDrop == null || canAcceptDrop(selected))
+                        return true;
+                }
+            }
+
+            return false;
+        }
+
+        void OnFoldoutToggle(ChangeEvent<bool> evt)
+        {
+            if (evt.previousValue != evt.newValue)
+            {
+                var isExpandedAction = new ChangeCategoryIsExpandedAction();
+                if (selection.Contains(this)) // expand all selected if the foldout is part of a selection
+                    isExpandedAction.categoryGuids = selection.OfType<SGBlackboardCategory>().Select(s => s.viewModel.associatedCategoryGuid).ToList();
+                else
+                    isExpandedAction.categoryGuids = new List<string>() { viewModel.associatedCategoryGuid };
+
+                isExpandedAction.isExpanded = evt.newValue;
+                isExpandedAction.editorPrefsBaseKey = blackboard.controller.editorPrefsBaseKey;
+                viewModel.requestModelChangeAction(isExpandedAction);
+            }
+        }
+
+        internal void TryDoFoldout(bool expand)
+        {
+            m_Foldout.SetValueWithoutNotify(expand);
+            if (!expand)
+                m_RowsContainer.RemoveFromHierarchy();
+            else
+                m_MainContainer.Add(m_RowsContainer);
+
+            var key = $"{blackboard.controller.editorPrefsBaseKey}.{viewDataKey}.{ChangeCategoryIsExpandedAction.kEditorPrefKey}";
+            EditorPrefs.SetBool(key, expand);
+        }
+
+        void OnMouseDownEvent(MouseDownEvent e)
+        {
+            if ((e.clickCount == 2) && e.button == (int)MouseButton.LeftMouse && IsRenamable())
+            {
+                OpenTextEditor();
+                e.PreventDefault();
+            }
+        }
+
+        internal void OpenTextEditor()
+        {
+            m_TextField.SetValueWithoutNotify(title);
+            m_TextField.style.display = DisplayStyle.Flex;
+            m_TitleLabel.visible = false;
+            m_TextField.Q(TextField.textInputUssName).Focus();
+            m_TextField.SelectAll();
+        }
+
+        void OnEditTextFinished()
+        {
+            m_TitleLabel.visible = true;
+            m_TextField.style.display = DisplayStyle.None;
+
+            if (title != m_TextField.text && String.IsNullOrWhiteSpace(m_TextField.text) == false && String.IsNullOrEmpty(m_TextField.text) == false)
```

Comment:
`IsNullOrWhiteSpace` should be sufficient here given the MSDN description:
```true if the value parameter is null or Empty, or if value consists exclusively of white-space characters.```

Reply:
sweet will remove that last bit then 👍